File: day2/day2.py
from functools import reduce

data = open('day2/day2.txt').read().split('\n')

def getBoxSurface(line):
    length, width, height = [int(d) for d in line.split('x')]
    areas = [length * width, length * height, width * height]
    return 2 * sum(areas) + min(areas)

def part1():
    return sum([getBoxSurface(l) for l in data])

def getRibbon(line):
    dims = sorted([int(d) for d in line.split('x')])
    return 2 * sum(dims[:-1]) + reduce((lambda x, y: x * y), dims)

def part2():
    return sum([getRibbon(l) for l in data])

print(f"Part 1: {part1()}")
print(f"Part 2: {part2()}")

File: day2/day2.js
const fs = require('fs');
const input = fs.readFileSync("day2/day2.txt", 'utf-8').split('\n');

class Box {
    constructor(line) {
        var items = line.split("x");
        this.length = +items[0];
        this.width = +items[1];
        this.height = +items[2];
        this.areas = [this.length * this.width, this.length * this.height, this.width * this.height];
        this.surface = 2 * this.areas.reduce((a, b) => a + b) + Math.min(...this.areas);
    }
}

function getBoxSurface(line) {
    var items = line.split("x");
    length = +items[0];
    width = +items[1];
    height = +items[2];
    areas = [length * width, length * height, width * height];
    return 2 * areas.reduce((a, b) => a + b) + Math.min(...areas);
}

function part1() {
    return input.map(l => getBoxSurface(l)).reduce((a, b) => a + b);
}

function getRibbon(line) {
    var items = line.split("x").map(d => +d);
    items.sort((a, b) => a - b);
    return items.reduce((a, b) => a * b) + (2 * (items[0] + items[1]))
}

function part2() {
    return input.map(l => getRibbon(l)).reduce((a, b) => a + b);
}

console.log("Part 1:" + part1());
console.log("Part 2:" + part2());